lib_eio/net.ml

```diff
@@ -267,3 +269,52 @@ let with_tcp_connect ?(timeout=Time.Timeout.none) ~host ~service t f =
   | exception (Exn.Io _ as ex) ->
     let bt = Printexc.get_raw_backtrace () in
     Exn.reraise_with_context ex bt "connecting to %S:%s" host service
+
+let accept_fork_domain ~on_error (domain_mgr, domains) (t : #listening_socket) connection_handler () =
+  Switch.run @@ fun sw ->
+  let flow, addr = accept ~sw t in
+  Fun.protect
+    ~finally:(fun () ->
+      Flow.close flow;
+      Semaphore.release domains
+    )
+    (fun () ->
+      Domain_manager.run domain_mgr (fun () ->
+        Semaphore.acquire domains ;
+        match connection_handler (flow :> stream_socket) addr with
+        | x -> x
+        | exception ex -> on_error (Exn.add_context ex "handling connection from %a" Sockaddr.pp addr)
+      )
+    )
+
+let run_server ?(max_connections=Int.max_int) ?shutdown ?(additional_domains) ?(on_error=raise) listening_socket connection_handler =
+  (if max_connections < 0 then invalid_arg "max_connections");
+  let connections = Semaphore.make max_connections in
+  let shutdown =
+    match shutdown with
+    | Some p -> p
+    | None -> fst (Promise.create ())
+  in
+  let connection_handler flow addr =
+    connection_handler flow addr ;
+    Semaphore.release connections ;
+  in
+  Switch.run @@ fun sw ->
+  let acceptor : unit -> unit =
+    match additional_domains with
+    | Some (domain_mgr, domains) ->
+      (if domains < 1 then invalid_arg "additional_domains");
+      let domains = Semaphore.make domains in
+      accept_fork_domain (domain_mgr, domains) listening_socket ~on_error connection_handler
+    | None -> (fun () -> accept_fork ~sw listening_socket ~on_error connection_handler)
+  in
+  let rec loop () =
+    Fiber.first
+      (fun () ->
+        Semaphore.acquire connections;
+        acceptor ();
+        loop ()
+      )
+      (fun () -> Promise.await shutdown )
```

Comment:
Hmm... Took me moment to understand this pattern.

If I understand correctly, then for each connection two fibers are created.  The second fiber merely waits for the shutdown signal and if that arrives, then the other fiber will be cancelled.

This makes me wonder, however, whether it would be possible to use something like a single switch for the cancellation of all the connections running under that switch.

Perhaps one could even remove the `shutdown` argument and instead have the whole server run under a switch that one can then cancel (e.g. by using `first` outside of the server).

Reply:
The idea with the current implementation is to gracefully stop the accepting loop, rather than (abruptly?) cancelling the currently running fibers. However, of course users can implement the cancel functionality as you suggested outside of eio.

Reply:
This system is not tail-recursive. Every new connection starts another fiber waiting for shutdown.

Instead, you can replace `?shutdown` with a `?sw` argument to let the connections outlive the accept loop if desired, and then use Eio's built-in cancellation.

Reply:
I have clarified the cancellation in documentation explicitly.